server/src/main/java/org/elasticsearch/cluster/routing/allocation/ShardsAvailabilityHealthIndicatorService.java

```diff
@@ -146,7 +255,133 @@ private static boolean isUnassignedDueToNewInitialization(ShardRouting routing)
+            SnapshotShardSizeInfo.EMPTY,
+            System.nanoTime()
+        );
+        allocation.setDebugMode(RoutingAllocation.DebugMode.ON);
+        ShardAllocationDecision shardAllocationDecision = allocationService.explainShardAllocation(shardRouting, allocation);
+        AllocateUnassignedDecision allocateDecision = shardAllocationDecision.getAllocateDecision();
+        LOGGER.trace(
+            "[{}]: Obtained decision: [{}/{}]",
+            shardRouting.shardId(),
+            allocateDecision.isDecisionTaken(),
+            allocateDecision.getAllocationDecision()
+        );
+        if (allocateDecision.isDecisionTaken() && AllocationDecision.NO == allocateDecision.getAllocationDecision()) {
+            if (LOGGER.isTraceEnabled()) {
+                LOGGER.trace(
+                    "[{}]: Working with decisions: [{}]",
+                    shardRouting.shardId(),
+                    allocateDecision.getNodeDecisions()
+                        .stream()
+                        .map(
+                            n -> n.getCanAllocateDecision()
+                                .getDecisions()
+                                .stream()
+                                .map(d -> d.label() + ": " + d.type())
+                                .collect(Collectors.toList())
+                        )
+                        .collect(Collectors.toList())
+                );
+            }
+            List<NodeAllocationResult> nodeAllocationResults = allocateDecision.getNodeDecisions();
+            boolean diagnosisFound = checkIsAllocationDisabled(diagnosisOutput, shardRouting, nodeAllocationResults);
+            IndexMetadata index = state.metadata().index(shardRouting.index());
+            if (index != null) {
+                diagnosisFound |= checkDataTierRelatedIssues(diagnosisOutput, index, shardRouting, nodeAllocationResults);
+            }
+            if (diagnosisFound == false) {
+                diagnosisOutput.accept(ACTION_CHECK_ALLOCATION_EXPLAIN_API, shardRouting);
+            }
+        }
+    }
+
+    private static Predicate<NodeAllocationResult> nodeHasDeciderResult(String deciderName, Decision.Type outcome) {
+        return (nodeResult) -> nodeResult.getCanAllocateDecision()
+            .getDecisions()
+            .stream()
+            .anyMatch(decision -> deciderName.equals(decision.label()) && outcome == decision.type());
+    }
+
+    private boolean checkIsAllocationDisabled(
+        BiConsumer<UserAction.Definition, ShardRouting> diagnosisOutput,
+        ShardRouting shardRouting,
+        List<NodeAllocationResult> nodeAllocationResults
+    ) {
+        if (nodeAllocationResults.stream().allMatch(nodeHasDeciderResult(EnableAllocationDecider.NAME, Decision.Type.NO))) {
+            diagnosisOutput.accept(ACTION_ENABLE_ALLOCATIONS, shardRouting);
+            return true;
+        }
+        return false;
+    }
+
+    private boolean checkDataTierRelatedIssues(
```

Comment:
Shall we document? 

Also for these private methods, should they be package default visible and unit tested? I like the tests you added already, however going through the indicator covers many of these methods in one unit, so if a unit test fails it's not immediately clear where to look to debug.
What do you think?

Reply:
I reworked the unit tests in [d75e324](https://github.com/elastic/elasticsearch/pull/85071/commits/d75e324e7af06e7931fc00316bd2c04616eb496d). I agree, they needed some simplification and focusing on just the diagnosis calls.